instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt

```diff
@@ -113,11 +116,149 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
+     * That was happening in legacy versions of NN.
+     */
+    @Test
+    fun additional_alternative_is_not_force_to_invoke_alternatives_observer() {
+        // Prepare with alternative routes.
+        setupMockRequestHandlers(coordinates)
+        val primaryRouteOnly = listOf(requestDirectionsRouteSync(coordinates).first())
+
+        // Play the slower alternative route.
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.startRecording()
+            mockLocationReplayerRule.playRoute(primaryRouteOnly.first())
+            mapboxNavigation.startTripSession()
+        }
+
+        // Wait for enhanced locations to start and then set the routes.
+        val firstLocationIdlingResource = FirstLocationIdlingResource(mapboxNavigation)
+        firstLocationIdlingResource.firstLocationSync()
+        runOnMainSync {
+            mapboxNavigation.setRoutes(primaryRouteOnly)
+        }
+
+        // The alternative route is missed, so we expect the route to be removed.
+        val firstAlternative = RouteAlternativesIdlingResource(mapboxNavigation)
+        firstAlternative.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("alternatives failed") {
+            Espresso.onIdle()
+        }
+        firstAlternative.unregister()
+
+        assertNotNull(firstAlternative.alternatives)
+
+        val externalAlternatives = DirectionsResponse.fromJson(
+            readRawFileText(activity, R.raw.route_response_alternative_continue)
+        ).routes().also {
+            assertEquals(1, it.size)
+        }
+
+        val nextAlternativeObserver = RouteAlternativesIdlingResource(mapboxNavigation)
+
+        var setRoutes: List<DirectionsRoute> = emptyList()
+        runOnMainSync {
+            setRoutes = mutableListOf(
+                mapboxNavigation.getRoutes().first()
+            ) + firstAlternative.alternatives!! + externalAlternatives
+
+            mapboxNavigation.setRoutes(setRoutes)
+        }
+
+        nextAlternativeObserver.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("next alternatives failed") {
+            Espresso.onIdle()
+        }
+        nextAlternativeObserver.unregister()
+
+        val countDownLatch = CountDownLatch(1)
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.stopRecording {
+                logE("history path=$it", "DEBUG")
+                countDownLatch.countDown()
+            }
+        }
+        countDownLatch.await()
+
+        // Verify alternative routes events were triggered.
+        assertNotNull(firstAlternative.routeProgress)
+        assertNotNull(firstAlternative.alternatives)
+
+        // Verify next alternative routes events were triggered without set routes
+        assertNotNull(nextAlternativeObserver.alternatives)
+        assertNotNull(nextAlternativeObserver.routeProgress)
+        assertTrue(setRoutes.size >= 2)
```

Comment:
You form that list in the test, is this really what you want to check?

Reply:
yes, it is. I wanted to verify logic, where `setRoutes` list is full enough (I don't need to control how many routes come to `AlternativesOberver`, I only want to check that it's full enought)

Reply:
Can we then check the size right after the creation? Here it's like hanging in the air. Quite confusing.

Reply:
sure, I'll move checks.